build.sh

```diff
@@ -340,127 +339,158 @@ if [[ "$DO_UPDATE" == true || ! -f "$HTML_CACHE/w3cbugs.csv" ]]; then
+  cp -p  quotes/out/cldr.inc "$HTML_CACHE"
+  if $VERBOSE; then
+    perl .pre-process-main.pl --verbose < "$HTML_SOURCE/$SOURCE_LOCATION" > "$HTML_TEMP/source-expanded-1"
   else
-    $QUIET || echo
-    $QUIET || echo "Local wattsi is not present; trying the build server..."
-
-    CURL_ARGS=( https://build.whatwg.org/wattsi \
-                --form "source=@$1" \
-                --form "caniuse=@$HTML_CACHE/caniuse.json" \
-                --form "w3cbugs=@$HTML_CACHE/w3cbugs.csv" \
-                --dump-header "$HTML_TEMP/wattsi-headers.txt" \
-                --output "$HTML_TEMP/wattsi-output.zip" )
-    if $VERBOSE; then
-      CURL_ARGS+=( --verbose )
-    elif $QUIET; then
-      CURL_ARGS+=( --silent )
+    perl .pre-process-main.pl < "$HTML_SOURCE/$SOURCE_LOCATION" > "$HTML_TEMP/source-expanded-1"
+  fi
+  perl .pre-process-annotate-attributes.pl < "$HTML_TEMP/source-expanded-1" > "$HTML_TEMP/source-expanded-2" # this one could be merged
+  perl .pre-process-tag-omission.pl < "$HTML_TEMP/source-expanded-2" | perl .pre-process-index-generator.pl > "$HTML_TEMP/source-whatwg-complete" # this one could be merged
+
+  function runWattsi {
+    # Input arguments: $1 is the file to run wattsi on, $2 is a directory for wattsi to write output
+    # to
+    # Output:
+    # - Sets global variable $WATTSI_RESULT to an exit code (or equivalent, for HTTP version)
+    # - $HTML_TEMP/wattsi-output directory will contain the output from wattsi on success
+    # - $HTML_TEMP/wattsi-output.txt will contain the output from wattsi, on both success and failure
+
+    rm -rf "$2"
+    mkdir "$2"
+
+    WATTSI_ARGS=()
+    if $QUIET; then
+      WATTSI_ARGS+=( --quiet )
     fi
-    curl "${CURL_ARGS[@]}"
-
-    # read exit code from the Wattsi-Exit-Code header and assume failure if not found
-    WATTSI_RESULT=1
-    while IFS=":" read -r NAME VALUE; do
-      if [ "$NAME" == "Wattsi-Exit-Code" ]; then
-        WATTSI_RESULT=$(echo "$VALUE" | tr -d ' \r\n')
-        break
+    WATTSI_ARGS+=( "$1" "$HTML_SHA" "$2" "$BUILD_TYPE" "$HTML_CACHE/caniuse.json" "$HTML_CACHE/w3cbugs.csv")
+    if hash wattsi 2>/dev/null; then
+      if [ "$(wattsi --version | cut -d' ' -f2)" -lt "$WATTSI_LATEST" ]; then
+        echo
+        echo "Warning: Your wattsi version is out of date. You should to rebuild an"
+        echo "up-to-date wattsi binary from the wattsi sources."
+        echo
       fi
-    done < "$HTML_TEMP/wattsi-headers.txt"
-
-    if [ "$WATTSI_RESULT" != "0" ]; then
-      mv "$HTML_TEMP/wattsi-output.zip" "$HTML_TEMP/wattsi-output.txt"
+      LOCAL_WATTSI=true
+      WATTSI_RESULT="0"
+      wattsi "${WATTSI_ARGS[@]}" || WATTSI_RESULT=$?
     else
-      UNZIP_ARGS=()
-      # Note: Don't use the -v flag; it doesn't work in combination with -d
-      if ! $VERBOSE; then
-        UNZIP_ARGS+=( -qq )
+      $QUIET || echo
+      $QUIET || echo "Local wattsi is not present; trying the build server..."
+
+      # TODO fix build server
```

Comment:
You'll need to pass it the git sha, right? You can assume a sha parameter. Should we also have a build type parameter, or should the server always do default?

Reply:
We'll also need a build type parameter. Unless the server is not used by Travis.